plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/ServerEvaluationStrategy.java

```diff
@@ -0,0 +1,197 @@
+     * Gets a map of all <strong>external</strong> addresses exposed by the container in the form of
+     * {@code "<address>:<port>"}
+     *
+     * @param containerInfo the ContainerInfo object that describes the container.
+     * @param internalAddress address passed into {@code getServers}; used as fallback if address cannot be
+     *        retrieved from containerInfo.
+     * @return a Map of port protocol (e.g. "4401/tcp") to address (e.g. "localhost:32317")
+     */
+    protected abstract Map<String, String> getExternalAddressesAndPorts(ContainerInfo containerInfo,
+                                                                        String internalAddress);
+
+    @Inject
+    protected ServerEvaluationStrategy(@Nullable @Named("che.docker.ip")          String internalAddressProperty,
+                                       @Nullable @Named("che.docker.ip.external") String externalAddressProperty) {
+        this.internalAddressProperty = internalAddressProperty;
+        this.externalAddressProperty = externalAddressProperty;
+    }
+
+    /**
+     * Constructs a map of {@link ServerImpl} from provided parameters, using selected strategy
+     * for evaluating addresses and ports.
+     *
+     * <p>Keys consist of port number and transport protocol (tcp or udp) separated by
+     * a forward slash (e.g. 8080/tcp)
+     *
+     * @param containerInfo the {@link ContainerInfo} describing the container.
+     * @param internalHost alternative hostname to use, if address cannot be obtained from containerInfo
+     * @param serverConfMap additional Map of {@link ServerConf}. Configurations here override those found
+     *        in containerInfo.
+     * @return a Map of the servers exposed by the container.
+     */
+    public Map<String, ServerImpl> getServers(ContainerInfo containerInfo,
+                                              String internalHost,
+                                              Map<String, ServerConfImpl> serverConfMap) {
+
+        Map<String, List<PortBinding>> portBindings;
+        Map<String, String> labels = Collections.emptyMap();;
+
+        if (containerInfo.getNetworkSettings() != null && containerInfo.getNetworkSettings().getPorts() != null) {
+            portBindings = containerInfo.getNetworkSettings().getPorts();
+        } else {
+            // If we can't get PortBindings, we can't return servers.
+            return Collections.emptyMap();
+        }
+        if (containerInfo.getConfig() != null && containerInfo.getConfig().getLabels() != null) {
+            labels = containerInfo.getConfig().getLabels();
+        }
+
+        Map<String, String> internalAddressesAndPorts = getInternalAddressesAndPorts(containerInfo, internalHost);
+        Map<String, String> externalAddressesAndPorts = getExternalAddressesAndPorts(containerInfo, internalHost);
+
+        Map<String, ServerImpl> servers = new LinkedHashMap<>();
+
+        for (String portProtocol : portBindings.keySet()) {
```

Comment:
Most of the getServers() logic is tested in DockerInstanceRuntimeInfoTest, as part of the tests there. I can move it out if that's preferred, but I think of ServerEvaluationStrategy as tightly coupled with DockerInstanceRuntimeInfo.